### PR TITLE
[READY FOR MERGE] Feature/add uds server auxiliary

### PR DESCRIPTION
**Pull request by sebclrsn**
PR opened for code review to discuss the final architecture:
- UdsBaseAuxiliary created as base class for UdsAuxiliary and UdsServerAuxiliary
- other modules have been moved to the new submodule udsaux.common
- ODX based callback configuration will be done in a follow up PR

The ODX parsing seems to be a lot of effort compared to the usage (python-uds only does the parsing for the client side, and even in that case the parameters names still need to be somehow known in order to define the request dictionary)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -39,7 +39,7 @@ PyYAML = "*"
 pylink-square = "^0.12.0"
 pyserial = "*"
 python-can = {extras = ["pcan"], version = "^4.0.0"}
-python-uds = {git = "https://github.com/BKaDamien/python-uds.git", rev = "2.0.2"}
+python-uds = {git = "https://github.com/sebclrsn/python-uds.git", branch = "feature/enable-server-side-isotp"}
```
Comment (sebastianpfischer) [inline pyproject.toml:42]: We now have the kiso-testing-python-uds, the changes should be put there

File: src/pykiso/lib/auxiliaries/udsaux/common/odx_parser.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+        """
+        root = ET.parse(self.odx_file)
+
+        self._xml_elements = {
+            child.attrib["ID"]: child
+            for child in root.iter()
+            if child.attrib.get("ID") is not None
+        }
+
+        for service_desc in self._xml_elements.values():
```
Comment (sebastianpfischer) [inline src/pykiso/lib/auxiliaries/udsaux/common/odx_parser.py:57]: Can you please comment the steps better? What is SDGS for, etc...

File: src/pykiso/lib/auxiliaries/udsaux/common/uds_callback.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            self.callback(aux, received_request)
+        else:
+            aux.send_response(self.response)
+
+    @staticmethod
+    def int_to_bytes(integer: int) -> bytes:
+        return integer.to_bytes((integer.bit_length() + 7) // 8, "big")
+
+
+def handle_data_download(
```
Comment (sebastianpfischer) [inline src/pykiso/lib/auxiliaries/udsaux/common/uds_callback.py:115]: I thought this was not specified by the protocol. Did I miss understood or is this a mistake?

**Review by sebastianpfischer — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 1 above)
Comment (sebclrsn) [reply pyproject.toml]: Updated to pykiso-python-uds ^2.1.0 ;)

File: src/pykiso/lib/auxiliaries/udsaux/common/odx_parser.py (hunk 2 above)
Comment (sebclrsn) [reply src/pykiso/lib/auxiliaries/udsaux/common/odx_parser.py]: As discussed I removed the OdxParser for the moment and now throw out a warning if the auxiliary is configured with an ODX file

**Comment by Pog3k** [comment]
Nice Feature, please add it to whats new section.

File: src/pykiso/lib/auxiliaries/udsaux/common/uds_callback.py (hunk 3 above)
Comment (sebclrsn) [reply src/pykiso/lib/auxiliaries/udsaux/common/uds_callback.py]: The DownloadData service is actually standardized by the ISO, so we can leave it here ;)

File: examples/test_uds/test_uds_server.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+import pykiso
+from pykiso.auxiliaries import uds_aux
+
+
+@pykiso.define_test_parameters(suite_id=1, case_id=1, aux_list=[uds_aux])
+class ExampleUdsServerTest(pykiso.BasicTest):
+    def setUp(self):
+        """Hook method from unittest in order to execute code before test case run."""
+        uds_aux.register_callback(0x1003, 0x50031234)
```
Comment (Pog3k) [inline examples/test_uds/test_uds_server.py:34]:
```suggestion
        uds_aux.register_callback(request = 0x1003, response =  0x50031234)
```

Comment (sebclrsn) [reply examples/test_uds/test_uds_server.py]: Should be good to go now, test example and docs have been extended as much as possible

File: docs/examples/uds_server_auxiliary.rst [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+                uds_aux.register_callback(callback)
+
+        def test_run(self):
+            """Actual test."""
+            ...
+
+        def tearDown(self):
+            """Unregister all callbacks from the external file."""
+            for callback in UDS_CALLBACKS:
+                uds_aux.register_callback(callback)
```
Comment (Pog3k) [inline docs/examples/uds_server_auxiliary.rst:218]:
```suggestion
                uds_aux.unregister_callback(callback)
```

File: docs/examples/uds_server_auxiliary.rst [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+
+.. code:: python
+
+    import pykiso
+    from pykiso.auxiliaries import uds_aux
+
+    # only used for type-hinting the custom callback
+    from pykiso.lib.auxiliaries.udsaux import UdsServerAuxiliary
+
+    def custom_callback(ecu_reset_request: list, aux: UdsServerAuxiliary) -> None:
```
Comment (Pog3k) [inline docs/examples/uds_server_auxiliary.rst:239]: where do you add this callback in this example ??

File: docs/examples/uds_server_auxiliary.rst [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+                extended_diag_session_callback.call_count,
+                0,
+                "Expected UDS request was not sent by the client after 10s",
+            )
+
+        def tearDown(self):
+            """Unregister all callbacks registered by the auxiliary."""
+
+            for callback in uds_aux.callbacks:
+                uds_aux.register_callback(callback)
```
Comment (Pog3k) [inline docs/examples/uds_server_auxiliary.rst:352]:
```suggestion
                uds_aux.unregister_callback(callback)
```

File: docs/examples/uds_server_auxiliary.rst [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+
+    import pykiso
+    from pykiso.auxiliaries import uds_aux
+
+    # only used for type-hinting the custom callback
+    from pykiso.lib.auxiliaries.udsaux import UdsServerAuxiliary
+
+    from uds_callback_definition import UDS_CALLBACKS
+
+    def custom_callback(ecu_reset_request: list, aux: UdsServerAuxiliary) -> None:
```
Comment (Pog3k) [inline docs/examples/uds_server_auxiliary.rst:302]: where does it get reigistered ? 
Can you give an example for ecu_reset_request ?

File: src/pykiso/lib/auxiliaries/udsaux/common/uds_base_auxiliary.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        com: CChannel,
+        config_ini_path: Union[Path, str],
+        odx_file_path: Optional[Union[Path, str]] = None,
+        request_id: Optional[int] = None,
+        response_id: Optional[int] = None,
+        **kwargs,
+    ):
+        """Initialize attributes.
+
+        :param com: communication channel connector.
```
Comment (Pog3k) [inline src/pykiso/lib/auxiliaries/udsaux/common/uds_base_auxiliary.py:53]: use same order as defined in the init method

Comment (sebclrsn) [reply src/pykiso/lib/auxiliaries/udsaux/common/uds_base_auxiliary.py:53]: Fixed

**Review by BKaDamien — APPROVED** [review]

**Pull request merged**